src/dns.c

```diff
@@ -2191,163 +2226,244 @@ hsk_dns_name_parse(
     }
   }
 
+  // If no pointers were used, resume from current offset
   if (ptr == 0)
     res = off;
 
-  if (noff == 0) {
-    if (name)
-      name[noff] = '.';
-    noff += 1;
-  }
-
-  if (name)
-    name[noff] = '\0';
-
+  // Move message read pointer in data stream
   *data_ += res;
   *data_len_ -= res;
 
+  // Return length written
   return noff;
 }
 
 static bool
 hsk_dns_name_serialize(
-  const char *name,
+  const uint8_t *name,
   uint8_t *data,
   int *len,
   hsk_dns_cmp_t *cmp
 ) {
+  // Offset of `name` input array being read
+  int noff = 0;
+  // Offset of `data` array being written
   int off = 0;
-  int pos = -1;
-  int ptr = -1;
-  int begin = 0;
-  size_t data_len = 256;
-  int size;
-  int i;
-  char *s;
 
-  for (s = (char *)name, i = 0; *s; s++, i++) {
-    if (name[i] == '.') {
-      if (i > 0 && name[i - 1] == '.') {
+  for (;;) {
+    uint8_t label = name[noff];
+
+    // End of name "."
+    if (label == 0x00) {
+      // Almost made it! Last byte exceeds max
+      if (1 + noff > HSK_DNS_MAX_NAME) {
         *len = off;
         return false;
       }
 
-      size = i - begin;
+      if (data)
+        data[off] = 0x00;
 
-      if (size > HSK_DNS_MAX_LABEL) {
-        *len = off;
-        return false;
-      }
+      off++;
+      *len = off;
+      return true;
+    }
+
+    // Label size can not exceed maximum
+    if (label > HSK_DNS_MAX_LABEL) {
+      *len = off;
+      return false;
+    }
+
+    // Label plus size byte can not extend full name past maximum
+    if (label + 1 + noff > HSK_DNS_MAX_NAME) {
+      *len = off;
+      return false;
+    }
+
+    // Label compression
+    if (cmp) {
+      // Check label map for remainder of name
+      size_t p = (size_t)hsk_map_get(&cmp->map, &name[noff]);
+
+      // Add remainder of name to map if it's not there already
+      if (p == 0) {
+        size_t o = (size_t)(&data[off] - cmp->msg);
 
-      if (data) {
-        if (off + 1 + size > data_len) {
+        if (o < (2 << 13))
+          hsk_map_set(&cmp->map, &name[noff], (void *)o);
+      } else {
+        // If the string was already in the map,
+        // point to it instead of repeating the string.
+        // Ensure enough room for 2-byte pointer, this should only
+        // be an issue if we are pointing to a 1-byte label...!
+        if (2 + noff > HSK_DNS_MAX_NAME) {
```

Comment:
same here, `noff + 2`

Reply:
thanks fixed in f5200ab1251950b897e8ef3f71f5a11737aeaa5a